pkg/scheduler/util/assumecache/assume_cache.go

```diff
@@ -0,0 +1,418 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package assumecache
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+
+	"k8s.io/klog/v2"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/tools/cache"
+)
+
+// AssumeCache is a cache on top of the informer that allows for updating
+// objects outside of informer events and also restoring the informer
+// cache's version of the object. Objects are assumed to be
+// Kubernetes API objects that are supported by [meta.Accessor].
+//
+// Objects can referenced via their key, with [cache.MetaNamespaceKeyFunc]
+// as key function.
+type AssumeCache interface {
+	// Assume updates the object in-memory only.
+	//
+	// The version of the object must be greater or equal to
+	// the current object, otherwise an error is returned.
+	//
+	// Storing an object with the same version is supported
+	// by the assume cache, but suffers from a race: if an
+	// update is received via the informer while such an
+	// object is assumed, it gets dropped in favor of the
+	// newer object from the apiserver.
```

Comment:
I think the volumebinding plugin is affected by this race.

https://github.com/kubernetes/kubernetes/blob/f4e246bc93ffb68b33ed67c7896c379efa4207e7/pkg/scheduler/framework/plugins/volumebinding/binder.go#L469

stores an object that is locally modified:

https://github.com/kubernetes/kubernetes/blob/f4e246bc93ffb68b33ed67c7896c379efa4207e7/staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go#L128-L152